src/snowflake/cli/api/commands/flags.py

```diff
@@ -499,70 +501,83 @@ def execution_identifier_argument(sf_object: str, example: str) -> typer.Argumen
+    }
 
-    def _callback(project_path: Optional[str]):
-        dm = DefinitionManager(project_path)
+    try:
+        dm = DefinitionManager(project_path, {CONTEXT_KEY: {"env": env_cli_overrides}})
         project_definition = dm.project_definition
         project_root = dm.project_root
+        template_context = dm.template_context
 
-        if not getattr(project_definition, project_name, None):
+        if project_name is not None and not getattr(
+            project_definition, project_name, None
+        ):
             raise NoProjectDefinitionError(
                 project_type=project_name, project_file=project_path
             )
 
-        cli_context_manager.set_project_definition(project_definition)
-        cli_context_manager.set_project_root(project_root)
-        cli_context_manager.set_template_context(dm.template_context)
-        return project_definition
-
-    if project_name == "native_app":
-        project_name_help = "Snowflake Native App"
-    elif project_name == "streamlit":
-        project_name_help = "Streamlit app"
+    except MissingConfiguration:
+        if is_optional:
+            project_definition = None
+            project_root = None
+            template_context = {
+                CONTEXT_KEY: {
+                    "env": ProjectEnvironment(
+                        default_env={}, override_env=env_cli_overrides
+                    )
+                }
+            }
+        else:
+            raise
+
+    cli_context_manager.set_project_definition(project_definition)
+    cli_context_manager.set_project_root(project_root)
+    cli_context_manager.set_template_context(template_context)
+
+
+def _get_project_long_name(project_short_name: Optional[str]) -> str:
+    if project_short_name is None:
+        return "Snowflake"
+
+    if project_short_name == "native_app":
+        project_long_name = "Snowflake Native App"
+    elif project_short_name == "streamlit":
+        project_long_name = "Streamlit app"
     else:
-        project_name_help = project_name.replace("_", " ").capitalize()
+        project_long_name = project_short_name.replace("_", " ").capitalize()
+
+    return f"the {project_long_name}"
+
+
+def project_definition_option(project_name: Optional[str], is_optional: bool):
+    def project_definition_callback(project_path: str) -> None:
+        cli_context_manager.set_project_path_arg(project_path)
+        register_pre_execute_command(
+            lambda: register_project_definition(project_name, is_optional)
+        )
 
     return typer.Option(
         None,
         "-p",
         "--project",
-        help=f"Path where the {project_name_help} project resides. "
+        help=f"Path where {_get_project_long_name(project_name)} project resides. "
         f"Defaults to current working directory.",
-        callback=_callback,
+        callback=_callback(lambda: project_definition_callback),
         show_default=False,
```

Comment:
I think there are two changes that would be beneficial and would simplify the implementation:
1. Getting rid of `project_name`, it won't be used in v2, will remove the "long name" mapping. Removing it means we need to add proper validation in all commands requiring part of project definition.
2. `is_optional` does not have to be part of `register_project_definition`, we can use it as input for `exists` in `type.Option`. If project is optional the file may not exists and that's fine. If project is required, the file must exists.

Today project definition handling is tightly coupled with "project type" and this is a source of now convoluted code (it was good idea some time ago).

My suggestion would be to first to the above changes in separate PR to decouple refactoring and new features. I'm happy to contribute.

Reply:
1. and 2. I do not want to add behavior changes on this.
Did you mean to add 1. and 2. in a separate PR? or move some of the current changes into a separate PR?

Reply:
We had an offline discussion. I will remove project_name altogether as a follow up PR to unblock this feature.